pkg/assembler/assembler.go

```diff
@@ -0,0 +1,274 @@
+package assembler
+
+import (
+	"github.com/alecthomas/participle/v2"
+	f "github.com/consensys/gnark-crypto/ecc/stark-curve/fp"
+)
+
+var parser *participle.Parser[CasmProgram] = participle.MustBuild[CasmProgram](
+	// mandatory lookahead to disambiguate between productions:
+	// expr -> [reg + n] + [reg + m] and
+	// expr -> [reg + n]
+	participle.UseLookahead(5),
+)
+
+func CasmToBytecode(code string) ([]*f.Element, error) {
+	casmAst, err := parser.ParseString("", code)
+	if err != nil {
+		return nil, err
+	}
+	return encodeCasmProgram(*casmAst)
+}
+
+//
+// Functions that visit the AST in order to encode the instructions
+//
+
+const (
+	// offsets
+	op0Offset = 16
+	op1Offset = 32
+
+	// flag values
+	dstRegBit         = 48
+	op0RegBit         = 49
+	op1ImmBit         = 50
+	op1FpBit          = 51
+	op1ApBit          = 52
+	resAddBit         = 53
+	resMulBit         = 54
+	pcJumpAbsBit      = 55
+	pcJumpRelBit      = 56
+	pcJnzBit          = 57
+	apAddBit          = 58
+	apAdd1Bit         = 59
+	opcodeCallBit     = 60
+	opcodeRetBit      = 61
+	opcodeAssertEqBit = 62
+
+	// default values
+	biasedZero     uint16 = 0x8000
+	biasedPlusOne  uint16 = 0x8001
+	biasedMinusOne uint16 = 0x7FFF
+	biasedMinusTwo uint16 = 0x7FFE
+)
+
+func encodeCasmProgram(casmAst CasmProgram) ([]*f.Element, error) {
```

Comment:
Rather than encoding into bytecode, can we encode into our internal representation using `vm.Instruction`? Then we can easily add all of the test cases mentioned in the whitepaper.

Reply:
Maybe to another internal representation of `vm.Instruction` yes, but it needs to be another, because the vm will use the assembler as well for testing purposes, and ciruclar dependencies can occurs. If reusing the `vm.Instruction`, then I think we can either move the assembler inside VM or move `Instruction`s to the Assembler. I personally like the latter more than the former, and we can discuss it.

 Also you would need to write this `Instruction` representation into bytecode. I didn't go for it because it meant double the work in the testing part, you need to check that both part of the pipelines are working correctly (assembly -> internal_repr) and (internal_repr -> bytecode). I think it is definetly something we can improve upon

Reply:
Also, I think at some later stages, if required  even to deprecate the current parser (LL(5)) which is also built the grammar in runtime and go for a compiled grammar and a (LALR(1)) parser, which are faster and more powerful